### DISCUSSION TITLE
Modern image format support

### DISCUSSION THREAD

**Discussion by m-overlund** [OpenMage/magento-lts#2013] [Ideas]
I think OpenMage should include support for modern image formats, like webP, out of the box.

Yireo has done some great work in this regard:
https://github.com/yireo-magento1/Yireo_Webp

Also in regards to improving the page speed of OpenMage.

**Comment by addison74** [comment]
I agree to add webp file format for images. The only issue is related to extension itself. It needs to be intensive tested before being merged into OpenMage and we cannot count on developer's help. There are a few open issues but I did not check them in detail.

**Reply by m-overlund** [reply]
I agree. When i tested on our setup, some images failed conversion, resulting in zero byte webP files = broken images.
So maybe some checks for CMYK or other things fails or is needed.

**Comment by justinbeaty** [comment]
Another solution that requires no changes to OpenMage is to have some script that automatically makes webp versions of images in the media directory at some interval using cron or other methods.

Then, depending on the webserver (nginx or apache) you can serve the appropriate image to the web browser based on headers. This is how I do it with my OpenMage store.

nginx example:

```
map $http_accept $webp_suffix {
    default   "";
    "~*webp"  ".webp";
}
server {
    # ...
    location ~* \.(png|jpg)$ {
        # Only GET is allowed here
        limit_except GET { deny all; }

        try_files $uri$webp_suffix $uri =404;
    }
}
```

Apache example: https://gist.github.com/seeekr/2415528

Note that the two work slightly differently, the nginx config expects "imagename.jpg.webp" where the apache example expects "imagename.webp". I prefer the former just in case two files had the same name, like "imagename.jpg" and "imagename.png"

**Reply by addison74** [reply]
Interesting approach. Many thanks for sharing.

**Comment by Flyingmana** [comment]
just a personal opinion, we dont need to include everything into the main project, some things are fine to have as optional modules.
I see this as one of this cases, as basically the current formats we support are working, and are not fully terrible, but there are numerous approaches to improve it. Also there are several image formats, and this now only talks about one.

That said, Iam not against it and there definitive should be a way/an option to support newer formats. I just think, a separate module is generally ok.

**Reply by justinbeaty** [reply]
I agree that a separate module is fine. I have only quickly looked at the module, and it includes jQuery, and also seems to run blocking JS that modifies the DOM, which is bad for pagespeed. Plus it looks like it prefers to call a binary on the server to do the conversion.

**Comment by luigifab** [comment]
Are you watching me? I started the update of own of my module to automagically generate webp images same day of your question.

**Reply by m-overlund** [reply]
yes... yes I am (:

**Reply by m-overlund** [reply]
Sounds great - would love to test it out!

**Reply by luigifab** [reply]
The final crash test is ending!

To test it at localhost, _tomorrow_, you will need to run (I still not updated my documentation, 3 minutes light late):
- `composer require luigifab/openmage-apijs` ([6.9.0+](https://github.com/luigifab/openmage-apijs))
- `composer require luigifab/openmage-maillog` ([5.4.0+](https://github.com/luigifab/openmage-maillog)).
- `git apply `[changes.diff](https://github.com/OpenMage/magento-lts/files/9056620/changes.diff.txt)

You also need to install: `python3-pil` and `python3-scour` (Pillow for jpg jpeg gif png webp images, Scour for svg images).

1/ Then, go to _System / Configuration / Dialogs boxes_, and set Yes everywhere in General.

![image](https://user-images.githubusercontent.com/31816829/177597979-1d861261-bb01-4858-b35d-acc33c463d7b.png)

2/ Then, go to _System / Configuration / Emails and wysiwyg directives_, and create a Picture directive, for example:

![image](https://user-images.githubusercontent.com/31816829/177598141-746461ff-e8f7-486a-9d41-e92c67f31427.png)

3/ Then, go to your home cms page, add the following test: `{{picture code="test" product="simple" attribute="image"}}` (update the sku "simple" with an existing product sku with images). Go to your frontend:

![image](https://user-images.githubusercontent.com/31816829/177599577-8171a3b9-baa0-4198-a346-19b84eb5bc76.png)

Yes this is not automatic, you will need to update all your templates.

**Reply by luigifab** [reply]
Like us, you will love the animated GIF/PNG/WEBP resizing!

**Comment by fballiano** [comment]
I suggest to use a CDN which can automatically generate web images and recompress all of the images to match google' standards. I use one for all my customers, a few bucks/months and we avoid having hunderds thousands converted images on our servers.

**Reply by m-overlund** [reply]
Anyone you'd recommend for OpenMage?

**Reply by fballiano** [reply]
I use BunnyCDN for all my customers

**Comment by yafully** [comment]
My website uses My website adopts requireJs and jquery no prototypeJs, webp, less, ngx pagespeed, etc, but it is based on OPENMAGE and runs very smooth on PHP7.2 MYSQL5.7. Thank you for your hard work. 

[Site Address](https://www.roroos.com)

**Comment by kiatng** [comment]
PR #3384 and #3403 merged!
